File: 241008/아름다운 수/beautiful-number.cpp
#include <iostream>
#include <vector>

using namespace std;

int N;
int answer = 0;
vector<int> vec;

void check(){
    int idx = 0;
    
    bool flag = true;

    while(idx < vec.size()){
        int standard = vec[idx];
        int cnt = 0;
        while(idx < vec.size() && standard == vec[idx]){
            cnt++;
            idx++;
        }

        if(standard != cnt || cnt % standard != 0){
            flag = false;
        }
    }
    if(flag) answer++;
}

void makeBeautifulNum(){
    if(vec.size() == N){
        check();
        return;
    }
    else{
        for(int i = 1; i <= 4; i++){
            vec.emplace_back(i);
            makeBeautifulNum();
            vec.pop_back();
        }
    }
}

int main() {
    cin >> N;
    makeBeautifulNum();
    cout << answer <<"\n";
    return 0;
}

File: 241008/아름다운 수/beautiful-number.py
import sys

answer = 0
n = int(sys.stdin.readline())

# 아름다운 수 인지 확인하는 함수
def check(num_list):
    global answer
    flag = True
    i = 0
    
    while i < len(num_list):
        standard = num_list[i]
        cnt = 0

        while i < len(num_list) and standard == num_list[i]:
            cnt += 1
            i += 1
        
        if cnt != standard:
            flag = False

    if flag:
        answer += 1
            

# back tracking
def makeBeautifulNum():
    global num

    if len(num) == n:
        check(num)
        return

    else:
        for i in range(1, 5):
            num.append(i)
            makeBeautifulNum()
            num.pop()

num = []
makeBeautifulNum()
print(answer)